### PR TITLE
[WiP] OpenCL support

### PR DESCRIPTION
**Pull request by sklam**
OpenCL support
- `numba.ocl` namespace
- OpenCL driver bindings
- code generation thought SPIR 1.2
- vectorize support

Only tested on Linux + AMD hardware

The lack of generic address space support is making it difficult for high-level programming.  Current implementation assume all array in global memory space.  The lack of inttoptr/ptrtoint also makes array with non contiguous layout (numpy order 'A') impossible.  These limitation poses a barrier to further implementation.

We can provide this as an experimental support.  OpenCL 2.0 will have generic address space and may make inttoptr/ptrtoint possible.  

TODO:
- [ ] clean up user interface
- [ ] math functions

### CODE REVIEW

**Comment by breznak** [comment]
this looks exciting @sklam :+1:  any progress?

**Comment by seibert** [comment]
This work is blocked pending support of the SPIR 2.0 standard (which is not yet finalized) in OpenCL runtimes.  We really need the support for generic pointer address spaces in order for this Numba target to integrate into the general compiler pipeline.

**Comment by ssorgatem** [comment]
Does this (https://appdevelopermagazine.com/2462/2015/3/4/Khronos-Group-Introduces-New-Vulkan-Hardware-Driver-API-and-SPIR-V-Intermediate-Language-Shared-by-Vulkan-and-OpenCL-2.1/) affect this work in any way?

**Comment by seibert** [comment]
I think it provides a path in the future for OpenCL support in Numba, but it will require the OpenCL 2.1 standard to be approved and implemented.  (Historically, it has taken a long time for OpenCL standards to be implemented broadly, even if we ignore NVIDIA, who is typically the slowest since they already have CUDA.)

We also require LLVM to add the necessary SPIR-V support, but it sounds like that is already actively being investigated by developers on the LLVM mailing list.

**Comment by seibert** [comment]
Can we pull the ocl-compiler branch from Siu's repo into Numba (to preserve this bit of history), but then close this PR?  We won't be coming back to OpenCL until we see how SPIR-V is implemented in OpenCL 2.1.

**Comment by sklam** [comment]
The work is preserved as https://github.com/numba/numba/tree/ocl-compiler

**Comment by ARF1** [comment]
@seibert I stumbled across this old issue. In various posts from 2015 you mentioned several blocking requirements. Do you have the time to write a short follow-up with the state of availability now (4 years later)?

I imagine, other numba users searching for "numba" and "OpenCL" will land here as well and would benefit from a short update. - Though I understand if you do not have the time...

**Comment by seibert** [comment]
What has changed most with OpenCL is that the current standard has moved from SPIR (which is an LLVM-based IR) to SPIR-V (which is not LLVM at all).

Someone could write a Numba target for OpenCL, but would need to confirm the following:

* Is there a reliable path to translate something like LLVM IR into SPIR-V?  Numba pipeline is pretty focused on emitting LLVM IR, so making it do something entirely different would require a significantly larger amount of work.

* Does SPIR-V work on a wide enough range of hardware / software OpenCL implementations to be interesting for users?  (Apple's dropped OpenCL, so macOS is out unless someone has a SPIR-V to Metal compiler.)  Most important is probably NVIDIA and AMD GPUs + x86 multicore CPU on Linux and Windows.  Note that we were frequently surprised by OpenCL bugs during past OpenCL development, so we're a bit cautious.

Basically, we're not actively looking at OpenCL support (with all the other things on our todo list), but if someone wants to work on this, we would try to figure out a clear target extension API so that such an extension could develop asynchronous to the core Numba repository.

**Comment by ARF1** [comment]
@seibert Thank you very much for taking the time to explain this so clearly.

**Comment by thipokKub** [comment]
Hello, it's 2 year since the last post. But I think now there might be a solution to convert LLVM IR code to SPIR-V using [KhronosGroup/SPIRV-LLVM-Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator). And about SPIR-V support, I think it might be possible now as multiple vendor also support Vulkan (both nVidia, and AMD)

I could be wrong though, because I'm not sure about the technical detail. If so please correct me

**Comment by gmarkall** [comment]
> we would try to figure out a clear target extension API so that such an extension could develop asynchronous to the core Numba repository.

A quick 2022 update: the target extension API is pretty much complete and the CUDA target uses it, so it should be possible to add out-of-tree support for an OpenCL backend to Numba from 0.56 onwards. Unfortunately it's not really documented, but one could follow the pattern adopted by the CUDA target - the main classes to follow the implementation of would be `CudaCodeLibrary`, `JITCUDACodegen`, `CUDABackend`, `CUDACompiler`, `CUDATarget`, and `CUDADispatcher`.
